cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+				ID: lb.ID,
+			})
+
+			if err != nil {
+				if strings.Contains(err.Error(), "cannot be found") {
+					return nil
+				}
+				return fmt.Errorf("error fetching the load balancer: %w", err)
+			}
+
+			if lb != nil && lb.ProvisioningStatus != nil && *lb.ProvisioningStatus != string(infrav1beta2.VPCLoadBalancerStateDeletePending) {
+				_, err = s.IBMVPCClient.DeleteLoadBalancer(&vpcv1.DeleteLoadBalancerOptions{
+					ID: lb.ID,
+				})
+				if err != nil {
+					s.Error(err, "error deleting the load balancer")
+					return err
+				}
+				s.Info("Load balancer successfully deleted")
+			}
+		}
+	}
+	return nil
+}
+
+// DeleteVPCSubnet deletes VPC subnet.
+func (s *PowerVSClusterScope) DeleteVPCSubnet() error {
+	if !s.deleteResource(Subnet) {
+		return nil
+	}
+
+	for _, subnet := range s.IBMPowerVSCluster.Status.VPCSubnet {
+		if subnet.ID == nil {
+			continue
+		}
+
+		net, _, err := s.IBMVPCClient.GetSubnet(&vpcv1.GetSubnetOptions{
+			ID: subnet.ID,
+		})
+
+		if err != nil {
+			if strings.Contains(err.Error(), "Subnet not found") {
+				return nil
+			}
+			return fmt.Errorf("error fetching the subnet: %w", err)
+		}
+
+		_, err = s.IBMVPCClient.DeleteSubnet(&vpcv1.DeleteSubnetOptions{
+			ID: net.ID,
+		})
+		if err != nil {
+			return fmt.Errorf("error deleting VPC subnet: %w", err)
+		}
+		s.Info("VPC subnet successfully deleted")
+	}
+	return nil
+}
+
+// DeleteVPC deletes VPC.
+func (s *PowerVSClusterScope) DeleteVPC() error {
+	if !s.deleteResource(VPC) {
+		return nil
+	}
+
+	if s.IBMPowerVSCluster.Status.VPC.ID != nil {
+		vpc, _, err := s.IBMVPCClient.GetVPC(&vpcv1.GetVPCOptions{
+			ID: s.IBMPowerVSCluster.Status.VPC.ID,
+		})
+
+		if err != nil {
+			if strings.Contains(err.Error(), "VPC not found") {
+				return nil
+			}
+			return fmt.Errorf("error fetching the VPC: %w", err)
+		}
+
+		_, err = s.IBMVPCClient.DeleteVPC(&vpcv1.DeleteVPCOptions{
+			ID: vpc.ID,
+		})
+		if err != nil {
+			return fmt.Errorf("error deleting VPC: %w", err)
+		}
+		s.Info("VPC successfully deleted")
+	}
+	return nil
+}
+
+// DeleteTransitGateway deletes transit gateway.
+func (s *PowerVSClusterScope) DeleteTransitGateway() error {
+	if !s.deleteResource(TransitGateway) {
+		return nil
+	}
+
+	if s.IBMPowerVSCluster.Status.TransitGateway.ID != nil {
+		tg, _, err := s.TransitGatewayClient.GetTransitGateway(&tgapiv1.GetTransitGatewayOptions{
+			ID: s.IBMPowerVSCluster.Status.TransitGateway.ID,
+		})
+
+		if err != nil {
+			if strings.Contains(err.Error(), "gateway was not found") {
+				return nil
+			}
+			return fmt.Errorf("error fetching the transit gateway: %w", err)
+		}
+
+		tgConnections, _, err := s.TransitGatewayClient.ListTransitGatewayConnections(&tgapiv1.ListTransitGatewayConnectionsOptions{
+			TransitGatewayID: tg.ID,
+		})
+		if err != nil {
+			return fmt.Errorf("error listing transit gateway connections: %w", err)
+		}
+
+		if tgConnections.Connections != nil && len(tgConnections.Connections) > 0 {
```

Comment:
this if is not required since Connections is not a pointer to the array, if the connection len is zero, it won't loop.